### PR TITLE
Replace bit field manipulation functions with macros

### PR DESCRIPTION
**Pull request by akiss77**
JerryScript-DCO-1.0-Signed-off-by: Akos Kiss akiss@inf.u-szeged.hu

### CODE REVIEW

**Comment by LaszloLango** [comment]
I have related work in https://github.com/LaszloLango/jerryscript/commit/1223b994c7970c2a237f9907f4b0a0786bc18d4f
There is much optimization possibility in `ecma_make_*_value` functions too.

**Comment by akiss77** [comment]
This patch achieves quite good code size reduction on arm targets (2-8K on release builds).

filename | master/text | master/data | pr/text | pr/data | gain/text | gain/data
---------|-------------|-------------|---------|---------|-----------|----------
release.linux-armv7lhf/jerry | 195943 | 52 | 187192 | 52 | 8751 | 0
release.linux-armv7lhf-cp/jerry | 142148 | 52 | 136555 | 52 | 5593 | 0
release.linux-armv7lhf-cp_minimal/jerry | 87957 | 0 | 85396 | 0 | 2561 | 0
release.linux-armv7lhf-mem_stats/jerry | 196539 | 52 | 188297 | 52 | 8242 | 0
release.linux-armv7lhf-mem_stress_test/jerry | 195575 | 52 | 187588 | 52 | 7987 | 0
release.linux-x86_64/jerry | 278590 | 61 | 277340 | 52 | 1250 | 9
release.linux-x86_64-cp/jerry | 210702 | 61 | 210010 | 52 | 692 | 9
release.linux-x86_64-cp_minimal/jerry | 132993 | 0 | 132442 | 0 | 551 | 0
release.linux-x86_64-mem_stats/jerry | 280071 | 61 | 278549 | 52 | 1522 | 9
release.linux-x86_64-mem_stress_test/jerry | 279031 | 61 | 277469 | 52 | 1562 | 9
release.mcu_stm32f3-cp.bin/jerry | 142748 | 72 | 140232 | 72 | 2516 | 0
release.mcu_stm32f3-cp_minimal.bin/jerry | 82436 | 20 | 80912 | 20 | 1524 | 0
release.mcu_stm32f4-cp.bin/jerry | 142748 | 72 | 140232 | 72 | 2516 | 0
release.mcu_stm32f4-cp_minimal.bin/jerry | 82436 | 20 | 80912 | 20 | 1524 | 0
debug.linux-armv7lhf/jerry | 734016 | 60 | 733252 | 60 | 764 | 0
debug.linux-armv7lhf-cp/jerry | 537776 | 60 | 537020 | 60 | 756 | 0
debug.linux-armv7lhf-cp_minimal/jerry | 368272 | 24 | 366996 | 24 | 1276 | 0
debug.linux-armv7lhf-mem_stats/jerry | 737516 | 60 | 736660 | 60 | 856 | 0
debug.linux-armv7lhf-mem_stress_test/jerry | 734064 | 60 | 733300 | 60 | 764 | 0
debug.linux-x86_64/jerry | 817371 | 68 | 816185 | 68 | 1186 | 0
debug.linux-x86_64-cp/jerry | 635223 | 68 | 633962 | 68 | 1261 | 0
debug.linux-x86_64-cp_minimal/jerry | 443502 | 32 | 442300 | 32 | 1202 | 0
debug.linux-x86_64-mem_stats/jerry | 821176 | 68 | 820710 | 68 | 466 | 0
debug.linux-x86_64-mem_stress_test/jerry | 817391 | 68 | 816205 | 68 | 1186 | 0
debug.mcu_stm32f4-cp.bin/jerry | 547212 | 76 | 546324 | 76 | 888 | 0
debug.mcu_stm32f4-cp_minimal.bin/jerry | 367396 | 44 | 366500 | 44 | 896 | 0

Performance on x86_64/linux also improved slightly, especially on the longer running test cases (crypto-md5.js and crypto-sha1.js).

```
                  Benchmark |                   Perf(+ is better)
                 3d-cube.js |   0.259s ->   0.262s :  -1.093%  (+-1.765%) : [≈]
     access-binary-trees.js |   0.137s ->   0.132s :  +3.519%  (+-1.882%) : [+]
         access-fannkuch.js |   0.719s ->   0.717s :  +0.255%  (+-1.278%) : [≈]
            access-nbody.js |   0.316s ->   0.307s :  +2.847%  (+-0.907%) : [+]
bitops-3bit-bits-in-byte.js |   0.205s ->   0.203s :  +0.893%  (+-2.276%) : [≈]
     bitops-bits-in-byte.js |   0.262s ->   0.269s :  -2.610%  (+-2.358%) : [-]
      bitops-bitwise-and.js |   0.308s ->   0.296s :  +3.794%  (+-1.580%) : [+]
   controlflow-recursive.js |   0.090s ->   0.090s :  -0.557%  (+-7.179%) : [≈]
              crypto-aes.js |   0.508s ->   0.494s :  +2.854%  (+-1.115%) : [+]
              crypto-md5.js |   3.048s ->   2.736s : +10.226%  (+-0.374%) : [+]
             crypto-sha1.js |   1.351s ->   1.213s : +10.178%  (+-0.343%) : [+]
       date-format-tofte.js |   0.214s ->   0.208s :  +2.802%  (+-1.420%) : [+]
       date-format-xparb.js |   0.100s ->   0.097s :  +2.663%  (+-4.337%) : [≈]
             math-cordic.js |   0.293s ->   0.293s :  -0.057%  (+-2.025%) : [≈]
       math-partial-sums.js |   0.160s ->   0.159s :  +1.143%  (+-3.549%) : [≈]
      math-spectral-norm.js |   0.148s ->   0.148s :  +0.337%  (+-2.330%) : [≈]
            string-fasta.js |   0.449s ->   0.430s :  +4.306%  (+-1.561%) : [+]
            Geometric mean: |                 Speed up: 2.501% (+-0.650%) : [+]
```

File: jerry-core/jrt/jrt-bit-fields.h [hunk 1]
```diff
@@ -16,7 +17,27 @@
...
-extern uint64_t __attr_const___ jrt_set_bit_field_value (uint64_t, uint64_t, size_t, size_t);
+/**
+ * Extract a bit-field.
+ *
+ * @param container container to extract bit-field from
+ * @param lsb least significant bit of the value to be extracted
+ * @param width width of the bit-field to be extracted
+ * @return bit-field's value
+ */
+#define jrt_extract_bit_field(container, lsb, width) \
```
Comment (LaszloLango) [inline jerry-core/jrt/jrt-bit-fields.h]: Use uppercase for name of macros
`#define JRT_EXTRACT_BIT_FIELD(container, lsb, width) \`

File: jerry-core/jrt/jrt-bit-fields.h [hunk 2]
```diff
@@ -16,7 +17,27 @@
...
+/**
+ * Set a bit-field.
+ *
+ * @param container container to insert bit-field to
+ * @param new_bit_field_value value of bit-field to insert
+ * @param lsb least significant bit of the value to be inserted
+ * @param width width of the bit-field to be inserted
+ * @return bit-field's value
+ */
+#define jrt_set_bit_field_value(container, new_bit_field_value, lsb, width) \
```
Comment (LaszloLango) [inline jerry-core/jrt/jrt-bit-fields.h]: ditto

File: jerry-core/jrt/jrt-bit-fields.c [hunk 3]
```diff
@@ -1,62 +0,0 @@
...
-  return (shifted_value & bit_field_mask);
-} /* jrt_extract_bit_field */
-
-/**
- * Extract a bit-field from the integer.
- *
- * @return bit-field's value
- */
-uint64_t __attr_const___
-jrt_set_bit_field_value (uint64_t container, /**< container to insert bit-field to */
```
Comment (zherczeg) [inline jerry-core/jrt/jrt-bit-fields.c]: Why could not make this inline (in release mode)?

Comment (akiss77) [inline jerry-core/jrt/jrt-bit-fields.c]: Inline declaration in this C file can only have effect if we have LTO. Macros work whatever compiler/linker or setting we use.

Moreover, another issue with the current function-based approach is that it always casts everything to `uint64_t`, even if we would like to work with a different type. And then we cast the returned `uint64_t` again back to our original type. That cast is not something a compiler can optimize away.

**Comment by akiss77** [comment]
Rebased and pushed. (Also: found a solution to get rid of not-so-nice type casts that were present in previous versions of this patch.)

File: jerry-core/jrt/jrt-bit-fields.h [hunk 4]
```diff
@@ -16,7 +17,28 @@
...
+/**
+ * Extract a bit-field.
+ *
+ * @param container container to extract bit-field from
+ * @param lsb least significant bit of the value to be extracted
+ * @param width width of the bit-field to be extracted
+ * @return bit-field's value
+ */
+#define JRT_EXTRACT_BIT_FIELD(container, lsb, width) \
+  (((container) >> lsb) & ((((__typeof__ (container)) 1) << (width)) - 1))
```
Comment (zherczeg) [inline jerry-core/jrt/jrt-bit-fields.h]: __typeof__ is std c99?

Comment (akiss77) [inline jerry-core/jrt/jrt-bit-fields.h]: Now, that you mention... Interestingly, the build system does not complain about it. I see two ways of dealing with this:
* leave it in the currently proposed form, or
* let the macros have one more argument requiring the type of the container explicitly.

This first approach is a bit less standard-conforming (but seems to be supported by all compiilers out there, at least by clang for sure), while the second one needs more maintenance if the type of the first argument changes (anywhere where the macro is used).

Opinions?

Comment (zherczeg) [inline jerry-core/jrt/jrt-bit-fields.h]: I would prefer the second, since we want to maintain stdc99 support.

**Comment by akiss77** [comment]
@zherczeg I've updated the patch as suggested. Does it look better?

File: jerry-core/jrt/jrt-bit-fields.h [hunk 5]
```diff
@@ -16,7 +17,29 @@
...
+ * Extract a bit-field.
+ *
+ * @param type type of container
+ * @param container container to extract bit-field from
+ * @param lsb least significant bit of the value to be extracted
+ * @param width width of the bit-field to be extracted
+ * @return bit-field's value
+ */
+#define JRT_EXTRACT_BIT_FIELD(type, container, lsb, width) \
+  (((container) >> lsb) & ((((type) 1) << (width)) - 1))
```
Comment (LaszloLango) [inline jerry-core/jrt/jrt-bit-fields.h]: Any reason why did you remove the following asserts?
```C++
JERRY_ASSERT (lsb < JERRY_BITSINBYTE * sizeof (uint64_t));
JERRY_ASSERT (width < JERRY_BITSINBYTE * sizeof (uint64_t));
JERRY_ASSERT ((lsb + width) <= JERRY_BITSINBYTE * sizeof (uint64_t));
```

File: jerry-core/rcs/rcs-records.c [hunk 6]
```diff
@@ -222,7 +222,7 @@ rcs_record_get_field (rcs_record_t *rec_p, /**< record */
   JERRY_ASSERT (field_pos + field_width <= RCS_DYN_STORAGE_LENGTH_UNIT * JERRY_BITSINBYTE);
 
   uint32_t value = *(uint32_t *) rec_p;
-  return (uint32_t) jrt_extract_bit_field (value, field_pos, field_width);
+  return JRT_EXTRACT_BIT_FIELD (uint32_t, value, field_pos, field_width);
```
Comment (LaszloLango) [inline jerry-core/rcs/rcs-records.c]:
```C++
return (uint32_t) JRT_EXTRACT_BIT_FIELD (rcs_record_t, rec_p, field_pos, field_width);
```
couldn't work here?

File: jerry-core/jrt/jrt-bit-fields.h [hunk 7]
```diff
@@ -16,7 +17,29 @@
...
+ *
+ * @param type type of container
+ * @param container container to insert bit-field to
+ * @param new_bit_field_value value of bit-field to insert
+ * @param lsb least significant bit of the value to be inserted
+ * @param width width of the bit-field to be inserted
+ * @return bit-field's value
+ */
+#define JRT_SET_BIT_FIELD_VALUE(type, container, new_bit_field_value, lsb, width) \
+  (((container) & ~(((((type) 1) << (width)) - 1) << (lsb))) | (((type) new_bit_field_value) << (lsb)))
```
Comment (LaszloLango) [inline jerry-core/jrt/jrt-bit-fields.h]:
```C++
(((container) & ~(((((type) 1u) << (width)) - 1u) << (lsb))) | (((type) new_bit_field_value) << (lsb)))
```

File: jerry-core/jrt/jrt-bit-fields.h (hunk 5 above)
Comment (LaszloLango) [inline jerry-core/jrt/jrt-bit-fields.h]:
```C++
(((container) >> lsb) & ((((type) 1u) << (width)) - 1u))
```

Comment (akiss77) [inline jerry-core/jrt/jrt-bit-fields.h]: I had two reasons to remove the asserts:
* I didn't know how to insert them in the macro (which is an expression only, not a block of statements), and
* during the first experiments with the patch the compiler *did* emit warnings (which were turned to errors) when `lsb` and/or `width` was 'incompatible' with the type of `container`, and I've considered them as good replacements of the asserts.

As for the `1u`s: in the first case, it is certainly superfluous, since `1` is cast to `type`. In the second case, it might also be better to type-cast it instead of using the hard-coded `1u`. IMHO.

File: jerry-core/rcs/rcs-records.c (hunk 6 above)
Comment (akiss77) [inline jerry-core/rcs/rcs-records.c]: Good point, I guess it could work.

File: jerry-core/jrt/jrt-bit-fields.h (hunk 7 above)
Comment (akiss77) [inline jerry-core/jrt/jrt-bit-fields.h]: Same as above?

Comment (LaszloLango) [inline jerry-core/jrt/jrt-bit-fields.h]: Ok, leave as is.

File: jerry-core/jrt/jrt-bit-fields.h (hunk 5 above)
Comment (LaszloLango) [inline jerry-core/jrt/jrt-bit-fields.h]: I think the `do { ... } while (0)` trick could work here too. But if the compiler emit warnings, then it's Ok for me. The debug version of jerry runs really slow, so we should use assertions where they are necessary.

```C++
#define JRT_EXTRACT_BIT_FIELD(type, container, lsb, width) \
do \
{ \
  JERRY_ASSERT (lsb < JERRY_BITSINBYTE * sizeof (type)); \
  JERRY_ASSERT (width < JERRY_BITSINBYTE * sizeof (type)); \
  JERRY_ASSERT ((lsb + width) <= JERRY_BITSINBYTE * sizeof (type)); \
  (((container) >> lsb) & ((((type) 1) << (width)) - 1)) \
} while (0)
```

File: jerry-core/rcs/rcs-records.c [hunk 8]
```diff
@@ -35,7 +35,7 @@ rcs_record_set_field (rcs_record_t *rec_p, /**< record */
   JERRY_ASSERT (field_pos + field_width <= RCS_DYN_STORAGE_LENGTH_UNIT * JERRY_BITSINBYTE);
 
   uint32_t prev_value = *(uint32_t *) rec_p;
-  *(uint32_t *) rec_p = (uint32_t) jrt_set_bit_field_value (prev_value, value, field_pos, field_width);
+  *(uint32_t *) rec_p = JRT_SET_BIT_FIELD_VALUE (uint32_t, prev_value, value, field_pos, field_width);
```
Comment (LaszloLango) [inline jerry-core/rcs/rcs-records.c]: https://github.com/Samsung/jerryscript/pull/913/files#r53786515
ditto

File: jerry-core/rcs/rcs-records.c (hunk 6 above)
Comment (akiss77) [inline jerry-core/rcs/rcs-records.c]: Sorry, I have to revert that. It cannot be done. There is something nasty going on around `rcs_record_t`. It is declared as `uint8_t` but according to rcs-records.h, it is always at least 32 bits (header) and then comes some payload of "arbitrary" length. That must be why it is type cast to `uint32_t`.

Comment (LaszloLango) [inline jerry-core/rcs/rcs-records.c]: I see. Thank you for checking.

**Comment by LaszloLango** [comment]
LGTM

**Comment by zherczeg** [comment]
In general functions are easier to debug but I have no strong opinion, because I also use a lot of macros. I did not find the RPi2 results, am I missing something?

**Comment by bzsolt** [comment]
RPi2 results
master (https://github.com/Samsung/jerryscript/commit/76b37f34d0df7fadbecf57ac52345c6b50631b54)

                      Benchmark |                  Perf(+ is better)
                     3d-cube.js |  2.627s ->   2.543s :  +3.173%  (+-3.506%) : [≈]
         access-binary-trees.js |  1.407s ->   1.370s :  +2.607%  (+-2.291%) : [+]
             access-fannkuch.js |  7.960s ->   7.967s :  -0.084%  (+-2.530%) : [≈]
                access-nbody.js |  3.070s ->   3.077s :  -0.217%  (+-1.078%) : [≈]
    bitops-3bit-bits-in-byte.js |  1.730s ->   1.723s :  +0.386%  (+-1.912%) : [≈]
         bitops-bits-in-byte.js |  2.653s ->   2.603s :  +1.884%  (+-3.494%) : [≈]
          bitops-bitwise-and.js |  3.680s ->   3.663s :  +0.453%  (+-3.587%) : [≈]
       controlflow-recursive.js |  0.890s ->   0.870s :  +2.247%  (+-0.000%) : [+]
                  crypto-aes.js |  4.730s ->   4.643s :  +1.832%  (+-2.769%) : [≈]
                  crypto-md5.js | 21.490s ->  21.847s :  -1.660%  (+-0.982%) : [-]
                 crypto-sha1.js |  9.750s ->   9.817s :  -0.684%  (+-3.406%) : [≈]
           date-format-tofte.js |  2.233s ->   2.223s :  +0.448%  (+-2.090%) : [≈]
           date-format-xparb.js |  1.087s ->   1.053s :  +3.068%  (+-4.240%) : [≈]
                 math-cordic.js |  2.977s ->   2.943s :  +1.120%  (+-2.463%) : [≈]
           math-partial-sums.js |  1.660s ->   1.670s :  -0.602%  (+-0.000%) : [-]
          math-spectral-norm.js |  1.503s ->   1.457s :  +3.104%  (+-3.064%) : [+]
                string-fasta.js |  4.537s ->   4.467s :  +1.543%  (+-3.251%) : [≈]
                Geometric mean: |                Speed up: 1.106% (+-0.661%) : [+]

Binary sizes (bytes): 197,240 -> 193,144

**Comment by zherczeg** [comment]
LGTM
